dart-lang/dartdoc
# Switch search functionality to custom handling

This switches the current `typeahead` handling to custom handling. The typeahead library had a lot of functionality we don't use, it isn't supported anymore, and it also required us to also pull in jquery.

This is a preliminary change and sticks with the same search functionality and UI. Follow-up PRs can introduce style changes as well as expanded search functionality. 

This PR accomplishes the following:
- Removes the typeahead library
- Removes jquery and rewrites our code relying on it
- Writes custom search handling based off our `index.json`
- Avoids querying for the `index.json` multiple times unnecessarily.
- Fixes https://github.com/dart-lang/dartdoc/issues/2483

**Note:** These changes don't support Internet Explorer. Testing previous functionality, it didn't before either, so nothing is really lost here.

**Breaking changes:**
- Removes jquery library
- Removes typeahead(and bloodhound as a result) library

**Testing instructions:**
1. First, clone this branch locally.
2. Serve some documentation such as the SDK docs: `pub run grinder serve-sdk-docs`
- Verify searching at least works in all of the fields
- Verify the / works to grab focus and you are still able to type a slash in the input field
- Verify the suggestions are visible and properly positioned
- Verify you can move up and down between the suggestions to move between them. Also verify the text in the input field switches to each suggestion's name, but if you don't select any and click away from the search field the typed value remains.
- Verify moving to a suggestion with your arrow keys and clicking enter opens that suggestion's page.
- Without using arrow keys, verify you can see a hint(such as forEach when you type for) behind your written text. Proceed to verify clicking `Tab` fills the search bar with that hint.
- Verify you can click a suggestion with the mouse to open the appropriate page.
- Regression test any other functionality you expect from the search fields(s).
